### ISSUE TITLE
rubi_integrate takes over 2min to import

### ISSUE THREAD

**Issue by certik** [sympy/sympy#13241]
Here is a test case with the current master (96074807d8d003dbd83198d82ee0d5b16390709d) using `isympy`:
```
In [1]: %time from sympy.integrals.rubi.rubi import rubi_integrate
CPU times: user 2min 7s, sys: 160 ms, total: 2min 7s
Wall time: 2min 7s
```

**Comment by certik** [comment]
@Upabjojr wrote this in #13243, and I moved his comment here:

> rubi_integrate takes over 2 min to import

I guess you are loading the algebraic rules only. When I loaded all rules it took more than one hour.

**Comment by certik** [comment]
What is the underlying cause of the slow imports?

@Upabjojr how can I choose what rules import?

I read the documentation:
```
Signature: rubi_integrate(expr, var, showsteps=False)
Docstring:
Rule based algorithm for integration. Integrates the expression by appling
transformation rules to the expression.

Returns `Integrate` if an expression cannot be integrated.

Parameters
==========
expr : intergand expression
var : variable of integration

Returns Integral object if unable to integrate.
```
and it does not say.

**Comment by Upabjojr** [comment]
> What is the underlying cause of the slow imports?

It depends on MatchPy as an external library, which creates a complex data structure that gets updated for each rule that is added.

> how can I choose what rules import?

The main problem is that the rule subgroups depend on one another. The algebraic rules appear to be to only independent ones. I suggested that we keep the subgroups in different files, I believe @parsoyaarihant is currently putting comments on the imports according to which rules to load.

**Comment by certik** [comment]
So matchpy seems too slow to construct so many rules. Is it that it's using
inefficient data structures or algorithms, or is it that Python is too
slow, and it needs to be implemented in C++?

**Comment by Upabjojr** [comment]
Well, MatchPy seems to use the more efficient data structures known to this day. I don't know exactly how it works to explain the reasons of its slowness, I believe there could be still improvements on our part.

The matching conditions are all defined as lambdas, I suspect that makes it somehow harder to optimize the matching algorithm.

**Comment by parsoyaarihant** [comment]
> The matching conditions are all defined as lambdas, I suspect that makes it somehow harder to optimize the matching algorithm.

Actually @wheerd suggested that if the `Constraints` for a rule are subdivided divided into smaller individual constraints, it will make the matcher faster because it can backtrack faster. I can implement only `FreeQ` constraint directly using `MatchPy`  directly.

**Comment by Upabjojr** [comment]
> and it needs to be implemented in C++?

We have discussed this, the idea is to generate a decision tree out of MatchPy. It actually works already, but the generated file is huge and exceeds some Python limits.

**Comment by Upabjojr** [comment]
> Actually @wheerd suggested that if the Constraints for a rule are subdivided divided into smaller individual constraints, it will make the matcher faster because it can backtrack faster. 

Some identical constraints definitions are repeated many times in the generated rules. Each lambda definition in Python will be stored as a separate object, even if the lambda are identical. Does this potentially affect the performance of MatchPy?

**Comment by parsoyaarihant** [comment]
> Some identical constraints definitions are repeated many times in the generated rules

Each `Patterns` needs to be defined with all the `Constraints` for that `Pattern`. I don't think there is any workaround this.  Probably some optimisations can be done in MatchPy regarding this.

**Comment by Upabjojr** [comment]
> Each Patterns needs to be defined with all the Constraints for that Pattern.

What I fear is that identical constraint checks keep getting repeated in MatchPy.

Mathematica's pattern matcher is definitely much faster, though their code is private so we cannot inspect their algorithm.

**Comment by certik** [comment]
I think the pattern matching is just an intermediate step anyway. It'd be
nice to load it faster, but ultimately if it's too big an effort, it might
not be worthy.

I think we should spend our effort in figuring out how to do the decision
tree.

You said the main problem is that the file is too large. There are two ways
out. Either split it into subroutines and multiple files, or save the
decision tree to a file and then just interpret it in Python.

**Comment by Upabjojr** [comment]
I have opened an issue on MatchPy:
https://github.com/HPAC/matchpy/issues/27

**Comment by Upabjojr** [comment]
> I think we should spend our effort in figuring out how to do the decision tree.

@parsoyaarihant has been in contact with the MatchPy developers, they added the support for the generation of the decision tree to ease our work (huge thanks to @wheerd ). Unfortunately there are some problems: for what I understand, the size of the generated program is excessive.

I believe we should keep working on the SymPy-MatchPy interaction.

**Comment by certik** [comment]
I see, your idea is that this is matchpy's problem and that our job is to
interface matchpy well, and then we just need to work on matchpy.

I like the division of work. As long as we can get a performing decision
tree out of it, one way or another.

**Comment by asmeurer** [comment]
Regardless of the cause, these things should not be done at import time. They should be done the first time the function is called. If it takes two minutes, maybe we should `warnings.warn()` that it is precomputing something.

**Comment by axkr** [comment]
I had the same problems importing the Rubi rules in the Symja project.

In the end I programmed a converter which creates internal Java rule files (IntRules...java files) from the Rubi rules:
https://github.com/axkr/symja_android_library/tree/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/integrate/rubi45

The same technic is used for other rules for functions like Sin, Cos,...
https://github.com/axkr/symja_android_library/tree/master/symja_android_library/matheclipse-core/src/main/java/org/matheclipse/core/reflection/system/rules

Probably the import performance could be improved further if it's possible to serialize the internal rule representation effectively  and use some methods to share expressions which are equal only once in the serialized binary file format.

**Comment by sylee957** [comment]
Now, it takes more than 10 minutes to import in my machine
`1881190908 function calls (1669252728 primitive calls) in 870.287 seconds`
It may be different for the systems though. But I think we should investigate if there are some regressions about this.

**Comment by Upabjojr** [comment]
> Now, it takes more than 10 minutes to import in my machine
> `1881190908 function calls (1669252728 primitive calls) in 870.287 seconds`
> It may be different for the systems though. But I think we should investigate if there are some regressions about this.

We are currently looking for a GSoC 2019 student to work on with RUBI. We plan to generate a decision tree in C++ out of the RUBI rules to speed things up.

Please advertise the possibility of working for RUBI on SymPy this summer.
